main.py

```diff
@@ -0,0 +1,128 @@
+
+import newrelic.agent
+newrelic.agent.initialize("newrelic.ini", "production")
+app = newrelic.agent.register_application(timeout=10.0)
+
+import os
+
+import logging
+from datetime import datetime
+
+import discord
+from discord.ext import commands, tasks
+from discord.ext.commands import DefaultHelpCommand
+from discord_slash import SlashCommand
+from dotenv import load_dotenv
+
+# logs data to the discord.log file, if this file doesn't exist at runtime it is created automatically
+from cogs.slash.demonstrator_tools import DemonstratorTools
+from cogs.slash.student_tools import StudentTools
+from cogs.slash.utility import Utility
+from helpers.messages.errors import message__custom__error__check_failure, \
+    message__custom__error__missing_required_argument, message__custom__error__command_not_found, \
+    message__custom__error__bad_argument, message__custom__error__rate_limited, message__custom__error__unknown_error
+
+logger = logging.getLogger('discord')
+logger.setLevel(logging.INFO)  # logging levels: NOTSET (all), DEBUG (bot interactions), INFO (bot connected etc)
+handler = logging.FileHandler(filename='discord.log', encoding='utf-8', mode='w')
+handler.setFormatter(logging.Formatter('%(asctime)s:%(levelname)s:%(name)s: %(message)s'))
+logger.addHandler(handler)
+
+# load the private discord token from .env file.
+load_dotenv()
+
+# Initialise the Bot object with an accessible help Command object
+helpCommand = DefaultHelpCommand()
+
+bot = commands.Bot(
+    command_prefix='dh:',
+    intents=discord.Intents.all()
+)
+slash = SlashCommand(bot, sync_commands=True)
+
+bot_metrics = {}
+current_status_index = 0
+
+
+@newrelic.agent.background_task(name='main.on_ready', group='Task')
+@bot.event
+async def on_ready() -> None:
+    """
+    Do something when the bot is ready to use.
+    """
+    get_bot_metrics.start()
+    status_readout_loop.start()
+    print(f'------------------------------------------------------------------------------'
+          f'\n|  as of {datetime.utcnow()}, {bot.user.name} is operational  |'
+          f'\n------------------------------------------------------------------------------')
+
+
+@newrelic.agent.background_task(name='main.get_bot_metrics', group='Task')
+@tasks.loop(hours=24)
+async def get_bot_metrics() -> None:
+    await bot.wait_until_ready()
```

Comment:
`async def get_bot_metrics() -> None` is responsible for doing a bunch of API to get info/metrics on the bot usage
stuff like counting the number of members per server so a *unique* total member count can be generated
because of the number of potential API calls that could be made, this is only updated once every 24 hours

then the section function, `async def status_readout_loop() -> None` is responsible for updating the bot's 'activity' using the data gathered by `async def get_bot_metrics() -> None` which is stored within a dictionary.

Reply:
that makes sense, can you add it as a docstring to the function so that other people can understand in the future?

Reply:
yee